images/virtualization-artifact/pkg/controller/vd/vd_controller.go

```diff
@@ -0,0 +1,98 @@
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vd/internal/condition"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/vd/internal/source"
+	"github.com/deckhouse/virtualization-controller/pkg/dvcr"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+)
+
+const (
+	ControllerName = "vd-controller"
+
+	ImporterPodVerbose    = "3"
+	ImporterPodPullPolicy = string(corev1.PullIfNotPresent)
+)
+
+type Condition interface {
+	Handle(ctx context.Context, vd *virtv2.VirtualDisk) error
+}
+
+func NewController(
+	ctx context.Context,
+	mgr manager.Manager,
+	clientset kubernetes.Interface,
+	log logr.Logger,
+	importerImage string,
+	uploaderImage string,
+	dvcr *dvcr.Settings,
+) (controller.Controller, error) {
+	stat := service.NewStatService()
+	importer := service.NewImporterService(dvcr, mgr.GetClient(), clientset, importerImage, ImporterPodPullPolicy, ImporterPodVerbose, ControllerName)
+	uploader := service.NewUploaderService(dvcr, mgr.GetClient(), clientset, uploaderImage, ImporterPodPullPolicy, ImporterPodVerbose, ControllerName)
+	disk := service.NewDiskService(mgr.GetClient(), clientset, dvcr)
+
+	blank := source.NewBlankDataSource(stat, disk)
+
+	sources := source.NewSources()
+	sources.Set(virtv2.DataSourceTypeHTTP, source.NewHTTPDataSource(stat, importer, disk, dvcr))
+	sources.Set(virtv2.DataSourceTypeContainerImage, source.NewRegistryDataSource(stat, importer, disk, dvcr))
+	sources.Set(virtv2.DataSourceTypeObjectRef, source.NewObjectRefDataSource(stat, disk, mgr.GetClient()))
+	sources.Set(virtv2.DataSourceTypeUpload, source.NewUploadDataSource(stat, uploader, disk, dvcr))
+
+	reconciler := NewReconciler(
+		mgr.GetClient(),
+		Wrap(condition.NewDatasourceReadyHandler(blank, sources)),
+		Wrap(condition.NewProvisioningPodStartedHandler(importer, uploader)),
+		Wrap(condition.NewUserUploadStartedHandler(uploader, stat)),
+		Wrap(condition.NewAttachedHandler()),
+		Wrap(condition.NewReadyHandler(disk)),
+		Wrap(condition.NewResizedHandler(disk)),
+		internal.NewLifeCycleHandler(blank, sources, mgr.GetClient()),
+		internal.NewDeletionHandlerHandler(sources),
+		internal.NewAttacheeHandler(mgr.GetClient()),
+	)
+
+	vdController, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+
+	err = reconciler.SetupController(ctx, mgr, vdController)
+	if err != nil {
+		return nil, err
+	}
+
+	//if err = builder.WebhookManagedBy(mgr).
+	//	For(&virtv2.VirtualDisk{}).
+	//	WithValidator(NewValidator()).
+	//	Complete(); err != nil {
+	//	return nil, err
+	//}
+
+	log.Info("Initialized VirtualDisk controller", "image", importerImage)
+
+	return vdController, nil
+}
+
+type Wrapper struct {
+	cond Condition
+}
+
+func (w Wrapper) Handle(ctx context.Context, vd *virtv2.VirtualDisk) (bool, error) {
+	return false, w.cond.Handle(ctx, vd)
+}
+
+func Wrap(cond Condition) *Wrapper {
```

Comment:
I suggest to make Wrap more meaningful, e.g. NewNonRequeueHandler.
But why this wrap is needed? Why condition handlers can't implement Handle interface?